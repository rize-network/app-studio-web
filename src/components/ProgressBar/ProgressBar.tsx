import React from 'react';
import { ProgressBarProps } from './ProgressBar/ProgressBar.props';
import ProgressBarView from './ProgressBar/ProgressBar.view';

/**
 * ProgressBar component displays completion status of a task or process.
 */
const ProgressBarComponent: React.FC<ProgressBarProps> = (props) => (
  <ProgressBarView {...props} />
);

export const ProgressBar = ProgressBarComponent;
