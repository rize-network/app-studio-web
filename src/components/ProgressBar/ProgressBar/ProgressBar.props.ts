import { ViewProps } from 'app-studio';

export interface ProgressBarStyles {
  container?: ViewProps;
  bar?: ViewProps;
}

export interface ProgressBarProps extends ViewProps {
  /**
   * Current progress value
   */
  value?: number;
  /**
   * Maximum progress value
   */
  max?: number;
  /**
   * Color of the filled portion
   */
  color?: string;
  /**
   * Background color of the track
   */
  backgroundColor?: string;
  /**
   * Height of the progress bar
   */
  height?: number | string;
  /**
   * Border radius for rounded corners
   */
  radius?: number | string;
  /**
   * Custom styles for container or bar
   */
  views?: ProgressBarStyles;
}
