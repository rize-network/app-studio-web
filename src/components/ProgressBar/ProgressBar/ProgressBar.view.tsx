import React from 'react';
import { View, useTheme } from 'app-studio';
import { ProgressBarProps } from './ProgressBar.props';

const ProgressBarView: React.FC<ProgressBarProps> = ({
  value = 0,
  max = 100,
  color = 'theme.primary',
  backgroundColor = 'color.gray.200',
  height = 8,
  radius = 4,
  views,
  themeMode: elementMode,
  ...props
}) => {
  const { getColor, themeMode } = useTheme();
  const currentMode = elementMode ? elementMode : themeMode;

  const percentage = Math.min(100, Math.max(0, (value / max) * 100));
  const trackColor = getColor(backgroundColor, { themeMode: currentMode });
  const barColor = getColor(color, { themeMode: currentMode });

  return (
    <View
      role="progressbar"
      aria-valuenow={value}
      aria-valuemin={0}
      aria-valuemax={max}
      width="100%"
      height={height}
      backgroundColor={trackColor}
      borderRadius={radius}
      overflow="hidden"
      {...views?.container}
      {...props}
    >
      <View
        width={`${percentage}%`}
        height="100%"
        backgroundColor={barColor}
        borderRadius={radius}
        {...views?.bar}
      />
    </View>
  );
};

export default ProgressBarView;
