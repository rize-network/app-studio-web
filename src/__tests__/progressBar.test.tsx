import React from 'react';
import renderer from 'react-test-renderer';
import { ProgressBar } from '../components/ProgressBar/ProgressBar';

import { cleanup, render } from '@testing-library/react';

afterEach(() => {
  cleanup();
});

test('should render ProgressBar component without crashing', () => {
  render(<ProgressBar value={50} />);
});

test('ProgressBar with value 50 should match snapshot', () => {
  const tree = renderer.create(<ProgressBar value={50} />).toJSON();
  expect(tree).toMatchSnapshot();
});
