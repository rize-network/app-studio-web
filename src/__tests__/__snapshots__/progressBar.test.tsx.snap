// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar with value 50 should match snapshot 1`] = `
<div
  aria-valuemax={100}
  aria-valuemin={0}
  aria-valuenow={50}
  className="wh-100 ht-8px brs-4px ow-hidden"
  role="progressbar"
>
  <div
    className="wh-50 ht-100 brs-4px"
  />
</div>
`;
